### PR TITLE
Adds the Lance emergency shuttle

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Adds the Lance emergency  shuttle. from TG, from https://github.com/tgstation/tgstation/pull/74024 An emergency shuttle that violently slams into the station to the chosen docking location.
Perfect for when crew is hunkered down in cargo / security / bridge / science, and can't reach evacuation, but needs a shuttle.
5000 credits.
Everything in the path of the shuttle will be gibbed / destroyed.
Makes lockers /  SSU's destroy contents on force qdel.
When crew order a shuttle, it gets locked in, so admins can't rob the crew of their funds (or break things, in the case of the lance.

Adds various procs used by the lance, be it to allow for better shuttle smashing, to allow stuff to happen before an emergency shuttle is loaded, or other.
A turf decal used for better squaring

## Why It's Good For The Game

More shuttles is good, and this is a very fun shuttle for serious situations where escape is not viable.
Also potentially useful to hijackers or antagonists, to move the shuttle elsewhere... for a pretty penny.

## Images of changes

image

## Testing

Lots of slaming this bad boy into the station.

## TODO:

- [x] deal with qwertodos.
- [x] better indication of where this is slamming
- [x] double check decals / holopad in bridge because I like that thematically

## Changelog
:cl: Cheshify, Qwertytoforty
add: Adds the Lance emergency shuttle. An emergency shuttle that violently slams into the station to the chosen docking location. Perfect for when crew is hunkered down in cargo / security / bridge / science, and can't reach evacuation, but needs a shuttle. 5000 credits.
/:cl:

### CODE REVIEW

**Comment by Coolrune206** [comment]
"When crew order a shuttle, it gets locked in, so admins can't rob the crew of their funds (or break things, in the case of the lance."
If true could use an admin button in the shuttle manip or elsewhere to disable shuttle buying, for those shifts when we do want to do buildable shuttles or a custom one or the like.

**Comment by Qwertytoforty** [comment]
> "When crew order a shuttle, it gets locked in, so admins can't rob the crew of their funds (or break things, in the case of the lance." If true could use an admin button in the shuttle manip or elsewhere to disable shuttle buying, for those shifts when we do want to do buildable shuttles or a custom one or the like.

You can set the same variable to 1 on SSshuttles to prevent crew ordering a shuttle

**Comment by 1080pCat** [comment]
> [ ] better indication of where this is slamming

hi, idea guy here. once designated, the docking point and devastation trail gets that new warning sprite(the one vetus uses but not encrypted) but resprited to be yellow and a little bit more in theme, tiled across where its going to absolutely obliterate, and red for where its going to dock?

**Comment by Henri215** [comment]
Update PR title if its done

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 1]
```diff
@@ -0,0 +1,119 @@
+/obj/item/lance_docking_generator
+	name = "Lance Docking Beacon"
+	desc = "A signaling device for the thingymabober"
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:3]: description is incomplete.

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 2]
```diff
@@ -0,0 +1,119 @@
+/obj/item/lance_docking_generator
+	name = "Lance Docking Beacon"
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:2]:
```suggestion
	name = "lance docking beacon"
```

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		emagged = TRUE
+		to_chat(user, "<span class='notice'>You emag [src], removing its docking safeties.</span>")
+		var/turf/T = get_turf(src)
+		do_sparks(5, FALSE, T)
+		playsound(T, "sparks", 50, TRUE, SHORT_RANGE_SOUND_EXTRARANGE)
+		return TRUE
+
+/obj/item/lance_docking_generator/attack_self(mob/living/user)
+	if(!is_station_level(user.z))
+		to_chat(user, "You'll want this to dock on the station.")
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:28]: span class

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+				to_chat(user, "Dangerous landing conditions, aborting!")
+				qdel(port, force = TRUE)
+				return
+
+	if(min_x <= TRANSITION_BORDER_WEST + 1 || max_x >= TRANSITION_BORDER_EAST - 1)
+		to_chat(user, "Docking space area too close to edge of sector, aborting!")
+		qdel(port, force = TRUE)
+		return
+	if(min_y <= TRANSITION_BORDER_SOUTH + 1 || max_y >= TRANSITION_BORDER_NORTH - 1)
+		to_chat(user, "Docking space area too close to edge of sector, aborting!")
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:62]: span class

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	for(var/turf/BT in L2)
+		for(var/obj/Ohno in BT.contents)
+			if((istype(Ohno, /obj/machinery/atmospherics/supermatter_crystal) || istype(Ohno, /obj/singularity)) && !emagged)
+				to_chat(user, "Dangerous landing conditions, aborting!")
+				qdel(port, force = TRUE)
+				return
+	port.register()
+
+	log_admin("[key_name(user)] created the lance docking location at [COORD(port)].")
+	to_chat(user, "Landing zone set. The signaller vanishes!")
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:84]: aaaaand these too

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	new /obj/structure/lance_beacon(get_turf(src))
+	qdel(src)
+
+/obj/structure/lance_beacon
+	name = "lance docking beacon"
+	desc = "A beacon that allows the Lance to carefully slam into the station at plasteel destroying speeds"
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:90]:
```suggestion
	desc = "A beacon that allows the Lance to carefully slam into the station at plasteel destroying speeds."
```

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 7]
```diff
@@ -0,0 +1,119 @@
+/obj/item/lance_docking_generator
+	name = "Lance Docking Beacon"
+	desc = "A signaling device for the thingymabober"
+	icon = 'icons/obj/device.dmi'
+	icon_state = "gangtool-red"
+	var/gps
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:6]: Not being used?

File: code/modules/shuttle/lance_docking_targeter.dm [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+/obj/structure/lance_beacon
+	name = "lance docking beacon"
+	desc = "A beacon that allows the Lance to carefully slam into the station at plasteel destroying speeds"
+	icon = 'icons/obj/singularity.dmi'
+	icon_state = "beacon1"
+	layer = MOB_LAYER - 0.2 //so people can't hide it and it's REALLY OBVIOUS
+	anchored = TRUE
+	resistance_flags = INDESTRUCTIBLE
+	var/gps
```
Comment (Henri215) [inline code/modules/shuttle/lance_docking_targeter.dm:96]: This one too

File: code/modules/shuttle/ripple.dm [hunk 9]
```diff
@@ -20,3 +20,19 @@
...
+/obj/effect/temp_visual/ripple/lance_crush
+	name = "collision lights"
+	desc = "Something is coming through hyperspace in a very unsafe way. You *really* do not want to be standing here."
+	icon = 'icons/turf/walls/hierophant_wall_temp.dmi'
+	icon_state = "hierophant_wall_temp-0"
+	base_icon_state = "hierophant_wall_temp"
+	smoothing_flags = SMOOTH_BITMASK
+	smoothing_groups = list(SMOOTH_GROUP_HIERO_WALL)
+	canSmoothWith = list(SMOOTH_GROUP_HIERO_WALL)
+	color = rgb(255,0,0)
```
Comment (Henri215) [inline code/modules/shuttle/ripple.dm:34]:
```suggestion
	color = rgb(255, 0, 0)
```

File: code/modules/shuttle/shuttle.dm [hunk 10]
```diff
@@ -633,6 +650,18 @@
 									"<span class='userdanger'>You feel an immense \
 									crushing pressure as the space around you ripples.</span>")
 					L.gib()
+			else if(lance_docking) //corrupt the child, destroy them all
```
Comment (Henri215) [inline code/modules/shuttle/shuttle.dm:653]: 🤣

**Comment by Henri215** [comment]
Map looking good so far

File: code/modules/shuttle/lance_docking_targeter.dm (hunk 1 above)
Comment (Qwertytoforty) [reply code/modules/shuttle/lance_docking_targeter.dm:3]: updated

File: code/modules/shuttle/lance_docking_targeter.dm (hunk 3 above)
Comment (Qwertytoforty) [reply code/modules/shuttle/lance_docking_targeter.dm:28]: spanned

File: code/modules/shuttle/lance_docking_targeter.dm (hunk 4 above)
Comment (Qwertytoforty) [reply code/modules/shuttle/lance_docking_targeter.dm:62]: done

File: code/modules/shuttle/lance_docking_targeter.dm (hunk 5 above)
Comment (Qwertytoforty) [reply code/modules/shuttle/lance_docking_targeter.dm:84]: done

File: code/modules/shuttle/lance_docking_targeter.dm (hunk 7 above)
Comment (Qwertytoforty) [reply code/modules/shuttle/lance_docking_targeter.dm:6]: oops, supposed to make the gps spawned inside it equal to that

**Comment by Rex9001** [comment]
Hi, you need to credit the original pr which can be found here: https://github.com/tgstation/tgstation/pull/74024 unless its on a whim that they look pretty identical

**Comment by Qwertytoforty** [comment]
> Hi, you need to credit the original pr which can be found here: [tgstation/tgstation#74024](https://github.com/tgstation/tgstation/pull/74024) unless its on a whim that they look pretty identical

Sure, appologies, did it on the past ones, forgot to do it here.

**Comment by Rex9001** [comment]
> > Hi, you need to credit the original pr which can be found here: [tgstation/tgstation#74024](https://github.com/tgstation/tgstation/pull/74024) unless its on a whim that they look pretty identical
> 
> 
> 
> Sure, appologies, did it on the past ones, forgot to do it here.

Its no issue mate, thanks for the fast response!

**Comment by Cheshify** [comment]
Let's GOOOOO
Love your take on this shuttle, hope it gets merged.

**Comment by Qwertytoforty** [comment]
> Let's GOOOOO Love your take on this shuttle, hope it gets merged.

Thanks, certainly a fun shuttle! The only thing I wish I could have is the big engines, but something with our shuttle code still isn't right, so the 3x3 engine can't be used when the shuttle rotates.

**Comment by Cheshify** [comment]
> > Let's GOOOOO Love your take on this shuttle, hope it gets merged.
> 
> Thanks, certainly a fun shuttle! The only thing I wish I could have is the big engines, but something with our shuttle code still isn't right, so the 3x3 engine can't be used when the shuttle rotates.

The engines are a little fucky on our end, I might borrow your concepts if I ever want to redo it here.

File: code/datums/shuttles.dm [hunk 11]
```diff
@@ -8,6 +8,10 @@
 	var/description
 	var/admin_notes
 
+
```
Comment (1080pCat) [inline code/datums/shuttles.dm:11]:
```suggestion

```

**Review by 1080pCat — APPROVED** [review]
The shield generators are.. spotty to say the least. Even when centered the second one hardly covers 6 tiles total, you'll probably need *another* custom shield generator that uses offsets.
image

Other than that though, i dont see any issues map side.

**Review by FunnyMan3595 — APPROVED** [review]

**Review by Burzah — APPROVED** [review]

**Pull request merged**